File: util/randSVD.lua
randSVD={}

function randSVD:svd(input,d)
	--randomized singular value decomposition
	--input: input matrix
	--d: the num of estimated singular values

	local trans=false
	local X=input
	if input:size(2)>input:size(1) then
		trans=true
		X=X:t()
	end

	local omega=torch.randn(X:size(2),d)
	local Q,R=torch.qr(X*omega)
	local B=Q:t()*X
	local u,s,v=torch.svd(B)
	u=Q*u
	--s is a vector and u,v are matrices
	--elements in s are sortes from largest to smallest

	if trans then
		return v,s,u
	else
		return u,s,v
	end
end

File: Python/util/randSVD.py
import numpy as np

def svd(input,d):
    trans=False
    if input.shape[1]>input.shape[0]:
        trans=True
        input=input.transpose()
    omega=np.random.random([input.shape[1],d])
    Q,R=np.linalg.qr(np.dot(input,omega))
    B=np.dot(Q.transpose(),input)
    U,s,V=np.linalg.svd(B)
    V=V[:,0:d]
    U=np.dot(Q,U)
    if trans:
        return V,s,U
    else:
        return U,s,V